### PR TITLE
Add refman for transactional sources

### PR DESCRIPTION
**Pull request by viliam-durina**

### CODE REVIEW

File: reference-manual/src/main/asciidoc/work-with-jet/configure-fault-tolerance.adoc [hunk 1]
```diff
@@ -42,6 +44,11 @@ In some other cases, however, duplicate processing of data items can
 have quite surprising consequences. There is more information about this
 in our <<pitfalls-alo, Jet Concepts>> chapter.
 
+Also note that duplicate event processing can occur only in case of a
```
Comment (mtopolnik) [inline reference-manual/src/main/asciidoc/work-with-jet/configure-fault-tolerance.adoc]: in _the_ case

File: reference-manual/src/main/asciidoc/source-sink-connectors/kafka.adoc [hunk 2]
```diff
@@ -42,17 +36,22 @@ The Kafka source supports snapshots. Upon each snapshot it saves the
...
-it among all the sink processors on that member. You can provide a
-mapping function that transforms the items the sink receives into
-`ProducerRecord`s.
+The sink provides exactly-once guarantee at the cost of using Kafka
+transactions: Jet commits the produced records after each snapshot is
+completed. This greatly increases the latency because consumers see the
+records only after they are committed.
+
+If you use _at-least-once_ guarantee, records a visible immediately, but
+in case of a failure some records could be produced duplicately. You can
```
Comment (mtopolnik) [inline reference-manual/src/main/asciidoc/source-sink-connectors/kafka.adoc]: in _the_ case

File: reference-manual/src/main/asciidoc/source-sink-connectors/overview.adoc [hunk 3]
```diff
@@ -24,45 +24,66 @@ Journal feature). You can also set up an `IMap`/`ICache` as a sink for
...
-== Is it Replayable?
+== Is it Fault-Tolerant?
 
-Most finite data sources are replayable because they come from
-persistent storage. You can easily replay the whole dataset. However, an
-infinite data source may be of such nature that it can be consumed only
-once. An example is the TCP socket connector. Such sources are bad at
-fault tolerance: if anything goes wrong during the computation, it
-cannot be retried.
+Can the connector work in a way that even in case of a job failure and
```
Comment (mtopolnik) [inline reference-manual/src/main/asciidoc/source-sink-connectors/overview.adoc]: in _the_ case

File: reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc [hunk 4]
```diff
@@ -63,11 +63,11 @@ member will write to different file names. You can achieve the effect of
...
 
-If the job is running in exactly-once mode, items will be written to
-temporary files (ending with a ".tmp" suffix). When a snapshot is
-completed, the file will be atomically renamed to remove this suffix.
-Thanks to the two-phase commit of the snapshot, exactly-once guarantee
-is provided for the sink.
+If the job is running in exactly-once mode, items are written to
+temporary files (ending with a ".tmp" suffix). When Jet commits a
+snapshot, it atomically renames the file to remove this suffix. Thanks
+to the two-phase commit of the snapshot the sink provides exactly-once
```
Comment (mtopolnik) [inline reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc]: _the_ exactly-once guarantee

File: reference-manual/src/main/asciidoc/work-with-jet/configure-fault-tolerance.adoc [hunk 5]
```diff
@@ -25,12 +25,12 @@ Hazelcast Jet supports exactly-once processing with multiple sources and
...
 details see the summary in <<overview-of-sources-and-sinks>>.
 
-If you configure Jet job for exactly-once but use a sink with
-at-least-once guarantee, after a job restart you may get duplicates in
-the output. As opposed to doubly processing an input item, this is more
-benign because it just means getting the exact same result twice. If you
-use a source with at-least-once guarantee, some events might be
-processed multiple times.
+If you configure Jet job with the exactly-once guarantee but use a sink
+with at-least-once guarantee, after a job restart you may get duplicates
```
Comment (mtopolnik) [inline reference-manual/src/main/asciidoc/work-with-jet/configure-fault-tolerance.adoc]: with _the_ at-least once

**Comment by mtopolnik** [comment]
verify

File: reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc [hunk 6]
```diff
@@ -46,6 +46,13 @@ after startup: both new files and new content appended to existing ones. Files
 must be updated in an append-only fashion; if the existing content changes, the
 behavior is undefined.
 
+Often, GUI text editors, even if you just add lines at the end, delete
+the whole file and recreate it. This gives undefined results. For
+testing it's best to use:
+```
+echo "foo" >> your_file
```
[the remark this answers is not in the document]
Comment (viliam-durina) [reply reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc]: This exact command works, surprisingly, on Windows :)

File: reference-manual/src/main/asciidoc/source-sink-connectors/overview.adoc [hunk 7]
```diff
@@ -96,16 +117,17 @@ aren't packaged with Jet.
 
 .Sources and Sinks
 |===
-|Resource|Javadoc|Sample|Unbounded?|Replayable?|Checkpointing?|Distributed?|Data Locality
+|Resource|Javadoc|Sample|Unbounded?|Guarantee|Distributed?|Data Locality
```
[the remark this answers is not in the document]
Comment (viliam-durina) [reply reference-manual/src/main/asciidoc/source-sink-connectors/overview.adoc]: Seems OK to me. What do you suggest?

File: reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (olukas) [reply reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc]: wow, really? ok then :+1:

File: reference-manual/src/main/asciidoc/source-sink-connectors/overview.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (olukas) [reply reference-manual/src/main/asciidoc/source-sink-connectors/overview.adoc]: Based on our discussion it is ok how it is now. I'm marking this as resolved.

File: reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtopolnik) [reply reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc]: `echo` without argument works differently on Windows, then it prints `echo is OFF` if i remember right :)

**Comment by viliam-durina** [comment]
@olukas 
> I'm missing any explanation of what transactional sources/sinks mean in general (we can mention that they have some external state etc.).

It's explained [here](https://github.com/hazelcast/hazelcast-jet/pull/1922/files#diff-a7c000353089c61dc98fbba5201e8c1dR50-R54).

>Configure exactly-once for File sink and JMS source is not straight-forward - it is not configured directly in Sinks.file or Sources.jmsQueue but in their builder variants. We should have a code snippet for them in related part of documentation (5.7.2. File Sink and 5.10.1. Using JMS as a Source).

You don't need to configure anything. It's enabled by default, if you enable guarantee for the job. You need extra code if you want to disable it just for some connector.

> In section for Kafka sink and exactly once
>-    we should mention that "the snapshot interval has to be shorter than the Kafka transaction timeout" and explain why.

It's in javadoc [here](https://github.com/hazelcast/hazelcast-jet/pull/1922/files#diff-1bb1e513166ddcff2b7e56a9f3b592faR62-R71).

>-    we can mention kafka property isolation.level=read_committed - without it also uncommitted items can be observed which can be confusing.

added

> we should mention that source with non-durable topic cannot be exactly-once

It's [there](https://github.com/hazelcast/hazelcast-jet/blob/master/hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/JmsSourceBuilder.java#L177-L181).

>it can be useful to mentioned that XA connection factory should not be used for source

There's no problem in doing that.

File: reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (viliam-durina) [reply reference-manual/src/main/asciidoc/source-sink-connectors/file-socket.adoc]: It also prints the quotes, weird...

**Comment by olukas** [comment]
@viliam-durina 
> 
> > Configure exactly-once for File sink and JMS source is not straight-forward - it is not configured directly in Sinks.file or Sources.jmsQueue but in their builder variants. We should have a code snippet for them in related part of documentation (5.7.2. File Sink and 5.10.1. Using JMS as a Source).
> 
> You don't need to configure anything. It's enabled by default, if you enable guarantee for the job. You need extra code if you want to disable it just for some connector.
> 
Sure, but this code snippet how to disable it can be still useful for users since it is configured in `builder` method.

The rest seems ok for me now, thanks for clarification. Let's keep it in javadoc in `4.0`.

**Comment by cangencer** [comment]
JMS document is still referring that it doesn't have any fault tolerance.

**Pull request merged**